langchain4j/src/test/java/dev/langchain4j/data/document/splitter/DocumentByParagraphSplitterTest.java

```diff
@@ -53,8 +53,8 @@ void should_split_into_segments_with_one_paragraph_per_segment(String separator)
         segments.forEach(segment ->
                 assertThat(segment.text().length()).isLessThanOrEqualTo(maxSegmentSize));
         assertThat(segments).containsExactly(
-                textSegment(firstParagraph, metadata("index", "0").put("document", "0")),
-                textSegment(secondParagraph, metadata("index", "1").put("document", "0"))
+                textSegment(firstParagraph, metadata("index", "0").put("document", "0").put("startIndex", "0")),
+                textSegment(secondParagraph, metadata("index", "1").put("document", "0").put("startIndex", "27"))
```

Comment:
Is this supposed to be a start index in the original document? Whitespaces seems to be ignored

Reply:
Yes, the start index should be the index of the first character as it appears in the original text. 

However (and this might break the whole idea of it) white space is sort of normalised to one character between all words because of how splitting and then buliding up parts works. Essentially, if you have the string `"some       text"` before the split, after the splitting and segmenting, it will appear like `"some text"`. 

But, as far as I have tested, the single whitespace between words is respected by the currnet implementation of the start index.

Reply:
Is there a way to respect the original formatting?

Reply:
This would be a change to the underlying splitting logic. This PR makes no logical change to the output, just adds some metadata. If you want the whitespace logic to change, I would suggest that gets done in a separate issue/PR?

Reply:
@bracl I meant to respect the original formatting only in the index calculation. No need to do any changes to the splitting logic

Reply:
This start_index logic works by keeping track of an index and moving it for two reasons:
1. a new segment is created, and the index increases
2. an overlap is created, and the index decreases

In 1. the amount of increase is determined by the size of the `part`. This part does not have the original whitespace of the input text. 
In 2. any overlap is generated from the `part`, which again, does not have the original whitespace from the input text.

Basically, that information is lost as soon as we perform `String[] parts = split(document.text());`

Reply:
Can't we take the first part of the new segment (or overlap, whatever comes first) and do something like `document.text().indexOf(firstPart)`?

Reply:
Doing some testing on this, and im not sure it holds when the original document has characters other than spaces. `\n \r \t` characters are lost in the splitting, so then `indexOf` is unable to find the place in the text

Reply:
I've put `System.out.println(">>" + document.text().indexOf(part));` on `HierarchicalDocumentSplitter.java:118` and it seems to work. (This should be called only for the first segment in the segment)

Reply:
I think unfortunately going down the route of using `.indexOf` doesnt hold in all cases. 

Consider following test
```
    @Test
    public void test_start_indexes_with_lots_of_whitespace() {
        {
            DocumentSplitter splitter = DocumentSplitters.recursive(15, 10);

            List<TextSegment> textSegments = splitter.split(Document.from("aaaa aaaa aaaa\t\r\r\r\t\taaaa\r\r\rbbbb aaaa"));
            assertThat(textSegments).containsExactly(
                    textSegment("aaaa aaaa aaaa", metadata("index", "0").put("document", "0").put("char_start_index", "0")),
                    textSegment("aaaa\r\r\rbbbb aaaa", metadata("index", "1").put("document", "0").put("char_start_index", "20")) // This 20 comes from manually counting `aaaa aaaa aaaa\t\r\r\r\t\t`. 5 + 5 + 4 + 6.
            );
        }
    }
```

Running this test against the code in the branch as it stands results in this
![image](https://github.com/user-attachments/assets/d883b066-9c85-4553-b40b-fb51cc071e26)

If I use `indexOf` on the whole segmentText, I get this
![image](https://github.com/user-attachments/assets/8ec7dc0c-2664-4844-91a7-c23c1a61a4e9)
We can observe that the StringBuilder is replacing the original text with the `joinSeparator` when we build the string.

If I update to start using `indexOf` to get the indexes of the first part in the chunk, then I get this
![image](https://github.com/user-attachments/assets/9d9682ad-0899-4bb5-8dc1-1d573690d0f1)
We are finding the repeated word from earlier in the document. 

So, `indexOf` has a second parameter `fromIndex` that can be used to mark a place to start searching in the string. This would help us with duplicate text, however, it means that we would have to be keeping track of the current index throughout the whole document text as a "where to start from"... which is exactly what the PR is currently doing.

Okay, so im now thinking about changing the logic to update the index based on the original text and not the output of the stringBuilders. The problem with this is that when we go into a recursive splitter, we dont retain any of the information when it returns. Ultimately, because the stringBuilder doesnt know the original splitting character and replaces it with the joinSeparator we have issues. 

I think maybe I should be looking at the python library to see what they are doing? But I also feel like the structure that they have is actually quite different...? 

Thoughts? I am still very keen on trying to find a solution that satisfies as much as possible - i just worry that we are going to have to compromise somewhere...

Reply:
Yes, python implementation is different.

Well, if there is no easy way to do it, let's then keep it as it is now? But will this work for you then? Since one can't really rely on this...

Reply:
One can rely on the numbers _**if**_ the text is white-space-normalised on input. Which I agree is not something to be relied upon. 

What kind of documentation or caveats would we have to put in place to introduce something like this. 

Also, I understand if this is too far away from something that is completely correct and therefore not something you are comfortable with merging. I dont want this to be forced through for the sake of forcing it through

Reply:
TBH I am not convinced this change as of now is ready to be merged, considering it does not work as expected.
If you feel like you will benefit from this feature, let's keep it, but add a flag (default=disabled) to explicitly enable it and document the known issues in javadoc.

Reply:
Yep, that sounds reasonable. We can make it explicitly opt-in. 

I will update